dm/syncer/validator_checkpoint.go

```diff
@@ -335,34 +305,106 @@ func (c *validatorPersistHelper) persist(tctx *tcontext.Context, loc binlog.Loca
 			args = append(args, []interface{}{
 				c.cfg.SourceID, sourceTable.Schema, sourceTable.Table, r.srcJob.Key,
 				targetTable.Schema, targetTable.Table,
-				srcDataStr, dstDataStr, r.tp, pb.ValidateErrorState_NewErr,
+				string(srcDataStr), string(dstDataStr), r.tp, pb.ValidateErrorState_NewErr,
 			})
 		}
 	}
-	// todo: performance issue when using insert on duplicate? https://asktug.com/t/topic/33147
-	// todo: will this transaction too big? but checkpoint & pending changes should be saved in one tx
-	var err error
+
+	return c.execQueriesWithRetry(tctx, queries, args)
+}
+
+func (c *validatorPersistHelper) persistPendingRows(tctx *tcontext.Context, rev int64) error {
+	count := int(c.validator.getAllPendingRowCount()) + 1
+	queries := make([]string, 0, count)
+	args := make([][]interface{}, 0, count)
+
+	// delete pending row of previous failed persist
```

Comment:
Is it previous failed and flushed rows? don't understand failed persist😵‍💫

Reply:
should be `delete pending rows left by previous failed persist`

Reply:
> should be `delete pending rows left by previous failed persist`

Plz rephrase the comment by this (p.s. `persist`'d better be changed to `persistence`)

Reply:
persist here means the method name, not some verb.